keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+      count: 5
+      whenTargetAchieved: continue
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+        - name: job-container
+          image: job-image
+          command: ["./sample"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+- If the job object's size reaches to limit of the etcd and
+the job controller can't store a correct value in `.status.completedIndexes`,
+we probably can not evaluate the SuccessPolicy correctly.
+
+- If we allow to set unlimited size of the value in `.spec.successPolicy.onSucceededPods.indexes`,
+we have a risk similar to [KEP-3850: Backoff Limits Per Index For Indexed Jobs](https://github.com/kubernetes/enhancements/tree/76dcd4f342cc0388feb085e685d4cc018ebe1dc9/keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs#risks-and-mitigations).
+So, we limit the size of `indexes` to 64Ki.
+
+## Design Details
+
+### Job API
+
+We extend the Job API to set different policies by which a Job can be declared as succeeded.
+
+```golang
+type JobSpec struct {
+	...
+	// Specifies the policy when the job can be declared as succeeded.
+	// If empty, the default behavior applies - the job is declared as succeeded
+	// only when the number of succeeded pods equals to the completions.
+	// This policy can not be specified more than one.
+	//
+	// This field is alpha-level. To use this field, you must enable the
+	// `JobSuccessPolicy` feature gate (disable by default).
+	// +optional
+	SuccessPolicy *SuccessPolicy
+}
+
+// SuccessPolicy describes when a Job can be declared as succeeded based on the success of some Indexes.
+type SuccessPolicy struct {
+	// Represents the requirement on the succeeded pods.
+	// +optional
+	OnSucceededPods *SuccessPolicyOnSucceededPods
+}
+
+// SuccessPolicyOnSucceededPods describes a Job can be succeeded based on succeeded pods.
+type SuccessPolicyOnSucceededPods struct {
+	// Specifies a set of required indexes when .spec.completionMode =
+	// "Indexed" in a text format.
+	// The job is declared as succeeded if a set of indexes are succeeded. 
+	// The list of indexes must come within 0 to `.spec.completions` and
+	// must not contain duplicates. At least one element is required. 
+	// The indexes are represented as decimal integers
+	// separated by commas. The numbers are listed in increasing order. Three or
+	// more consecutive numbers are compressed and represented by the first and
+	// last element of the series, separated by a hyphen.
+	// For example, if the completed indexes are 1, 3, 4, 5 and 7, they are
+	// represented as "1,3-5,7".
+	// 
+	// +optional
+	Indexes string
+
+	// Specifies the required number of indexes when .spec.completionMode =
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
```

Comment:
As I described in the `Design Details` section, we only support `indexes` OR `count` fields. As you said, disallowing to set both of them would be better.

> exactly one of the fields `indexes` and `count` is specified for a requirement.

I will update the above sentence so that we can clarify that `indexes` and `count` are exclusive.

Reply:
Oh, I see thanks. Still, it makes sense to also mention that in API comments, because they are used to generate the user-facing documentation. For example, we do something similar with OnExitCodes and OnPodConditions, which are mutually exclusive [here](https://github.com/kubernetes/kubernetes/blob/909faa3a9b33161992af0684c5582353b5be6d9b/pkg/apis/batch/types.go#L217). The API [docs](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/job-v1/#beta-level)

Reply:
Thank you for the navigation.
I agree with pointing out the exclusiveness as an API comment.

Reply:
As I explained here: https://github.com/kubernetes/enhancements/pull/4062#discussion_r1460681741, we have a story that `indexes` and `count` are not mutually exclusive.